### PR TITLE
feedback implementation

### PR DESCRIPTION
**Pull request by wafalakhal**
- Implement feedback API
- Add UI view for sending feedback

### CODE REVIEW

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/SettingsActivity.java [hunk 1]
```diff
@@ -44,10 +48,6 @@ protected void onCreate(Bundle savedInstanceState) {
         binding.toolbar.setStartTextTitle(getResources().getString(R.string.settings_app_bar));
         setSupportActionBar(binding.toolbar);
 
-        binding.toolbar.setBackClickListener(() -> {
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/SettingsActivity.java:47]: Why was this deleted?

Comment (wafalakhal) [reply mobile/src/main/java/rs/readahead/washington/mobile/views/activity/SettingsActivity.java:47]: thank you djordje, this should not be deleted, I will add it

File: mobile/src/main/java/rs/readahead/washington/mobile/data/feedback/remote/FeedbackApiService.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import rs.readahead.washington.mobile.data.entity.feedback.FeedbackPostResponse
+import rs.readahead.washington.mobile.data.feedback.utils.ParamsNetwork
+
+@JvmSuppressWildcards
+interface FeedbackApiService {
+
+    @POST
+    fun submitFeedback(
+        @Url
+        url: String = "https://api.feedback.tella-app.org/opinions/",
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/data/feedback/remote/FeedbackApiService.kt:18]: can you put the Feedback Url in the `ParamsNetwork` constants?

File: mobile/src/main/java/rs/readahead/washington/mobile/data/feedback/repository/FeedbackRepositoryImp.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import rs.readahead.washington.mobile.domain.entity.feedback.FeedbackStatus
+import rs.readahead.washington.mobile.domain.exception.NoConnectivityException
+import rs.readahead.washington.mobile.domain.repository.feedback.FeedBackRepository
+import javax.inject.Inject
+
+class FeedbackRepositoryImp @Inject internal constructor(
+        private val apiService: FeedbackApiService, private val dataSource: DataSource) : FeedBackRepository {
+    companion object {
+        private const val TELLA_PLATFORM = "wearehorizontal"
+    }
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/data/feedback/repository/FeedbackRepositoryImp.kt:21]: Don't you think we should put this in the `NetworkParams` constants?

File: mobile/src/main/java/rs/readahead/washington/mobile/data/feedback/repository/FeedbackRepositoryImp.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    override fun submitFeedback(feedbackBody: FeedbackBodyEntity): Single<FeedbackPostResult> {
+        return apiService.submitFeedback(
+                data = feedbackBody,
+                tellaPlatform = TELLA_PLATFORM,
+        ).map { it.mapToDomainModel() }
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/data/feedback/repository/FeedbackRepositoryImp.kt:27]: Consider renaming the lambda to `response` instead of it.

Comment (ahlem-jarrar) [reply mobile/src/main/java/rs/readahead/washington/mobile/data/feedback/repository/FeedbackRepositoryImp.kt:27]: also you should handle the error too.

File: mobile/src/main/java/rs/readahead/washington/mobile/data/feedback/repository/FeedbackRepositoryImp.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+                            .subscribeOn(Schedulers.io()).subscribe({
+                            }, { throwable ->
+                                throwable.printStackTrace()
+                            })
+                }
+    }
+
+    override fun displayFeedbackSent(): Single<Boolean> {
+        return Single.fromCallable {
+            true
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/data/feedback/repository/FeedbackRepositoryImp.kt:46]: why is this always sending true?

**Review by ahlem-jarrar — CHANGES_REQUESTED** [review]
I still need to finish the review, but you can start with this while I finish the review, also please make sure to add some unit testing and documentation to the feature!

File: mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/feedback/SendFeedbackViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+    }
+
+    fun getFeedBackDraft() {
+        disposables.add(dataSource.getFeedbackDraft().subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).doOnSubscribe { _progress.postValue(true) }.doFinally { _progress.postValue(false) }.subscribe({ draft ->
+            _draftFeedBackInstance.postValue(draft)
+
+        }) { throwable: Throwable? ->
+            Timber.d(throwable)
+            FirebaseCrashlytics.getInstance().recordException(throwable!!)
```
[the remark this answers is not in the document]
Comment (wafalakhal) [reply mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/feedback/SendFeedbackViewModel.kt:102]: I don't think so because it is very rare and we don't have anything about it on figma. I asked dhekra and she didn't handle it but I can ask Caro too about it.

File: shared-ui/src/main/java/org/hzontal/shared_ui/utils/DialogUtils.kt [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        overlay.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
+        frameLayout.addView(overlay)
+
+        // Create a transparent view to disable interaction with the views behind
+        val disableView = View(context)
+        disableView.setBackgroundColor(ContextCompat.getColor(context, android.R.color.transparent))
+        disableView.isClickable = true
+        disableView.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
+        frameLayout.addView(disableView)
+        val binding = LayoutBottomMessageWithButtonBinding.inflate(LayoutInflater.from(context), container, false)
```
Comment (djordjemarusic) [inline shared-ui/src/main/java/org/hzontal/shared_ui/utils/DialogUtils.kt:63]: Yellow text against white background : In snackbars the yellow button text should change from the primary yellow (#D6933B) to "yellow/darker" (#8B4F00)

File: mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/feedback/SendFeedbackViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        disposables.add(feedbackRepository.submitFeedbackInstance(instance).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).doOnSubscribe { _progress.postValue(true) }.doFinally { _progress.postValue(false) }.subscribe({
+            it.apply {
+                _feedbackSubmitted.postValue(true)
+            }
+        }) { throwable: Throwable? ->
+            Timber.d(throwable)
+            if (throwable is NoConnectivityException) {
+                _feedbackSubmitted.postValue(false)
+            } else {
+                FirebaseCrashlytics.getInstance().recordException(throwable!!)
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/feedback/SendFeedbackViewModel.kt:58]: I think we should submit a value for all the exceptions not only for `NoConnectivityException`, you can make an enum to identify the exception type and a work based on that, make sure to apply the logic if need in another place please.

File: mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/feedback/SendFeedbackViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import rs.readahead.washington.mobile.data.database.DataSource
+import rs.readahead.washington.mobile.domain.entity.feedback.FeedbackInstance
+import rs.readahead.washington.mobile.domain.exception.NoConnectivityException
+import rs.readahead.washington.mobile.domain.repository.feedback.FeedBackRepository
+import timber.log.Timber
+import javax.inject.Inject
+
+
+@HiltViewModel
+class SendFeedbackViewModel @Inject constructor(private val feedbackRepository: FeedBackRepository, private val dataSource: DataSource) : ViewModel() {
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/feedback/SendFeedbackViewModel.kt:22]: I don't understand why you are injecting dataSource here and inside the repository.

**Review by djordjemarusic — APPROVED** [review]

**Review by ahlem-jarrar — APPROVED** [review]
